### ISSUE TITLE
RC 3.9.1 Sidebar user status wrong and showing "@" symbols

### ISSUE THREAD

**Issue by TheWrongGuy** [RocketChat/Rocket.Chat#19809]
### Description:

We had wrong user statuses in the sidebar for a long time, some people were showed offline all the time. Because this was addressed in issue #19209 and should be fixed in version 3.8.0, I was thrilled to upgrade to the latest version 3.9.1. However the sidebar user status seems to be broken still. For many people it doesn't show the correct status and instead displays a "@" symbol. The status is correct in the user info window, so at some point the status seems to be recognized correctly.

### Steps to reproduce:

1. Klick on a direct messages chat
2. Click on user info symbol
3. The status is correct in user info window, but wrong in top bar and in the sidebar showing a "@"

### Expected behavior:

The status in sidebar and top bar should match the correct status in the user info window.

### Actual behavior:

<img width="1003" alt="UserStatusWrongSidebar" src="https://user-images.githubusercontent.com/72255550/101601117-93413c80-39fc-11eb-93e3-40ec89ec9a11.png">

### Server Setup Information:

- Version of Rocket.Chat Server: 3.9.1
- Operating System: CentOS 7.8
- Deployment Method: Manual
- Number of Running Instances: 1
- DB Replicaset Oplog: /
- NodeJS Version: 12.8.4
- MongoDB Version: 4.0.19

### Client Setup Information

- Desktop App or Browser Version: MacOS Client 3.0.7 / Google Chrome Webbrowser
- Operating System: MacOS Big Sur

**Comment by ankar84** [comment]
Interesting, but in my case on 3.9.1 all seem to be perfect in part of users statuses
![image](https://user-images.githubusercontent.com/4023037/101609998-af52d700-3a18-11eb-8cd8-293dcbc87b8c.png)

**Comment by TheWrongGuy** [comment]
Does somebody have an idea for checking a specific log or triggering one to maybe track down this issue? For some direct messages conversations the status works fine for me.

**Comment by TheWrongGuy** [comment]
It seems to be related with old direct messages conversations. I created a new test account and wrote a message to my actual account and the status is fine. It feels like a leftover from a previous issue or version stuck somewhere. Is it possible to renew the conversation somehow? Or maybe export the content delete the conversation and reimport it? It would like to find a solution without data loss.

**Comment by Gummikavalier** [comment]
We haven't seen the issue, but we can see @name in search-field when typing a name that one does not have contacted earlier. So the symbol itself seems to be coded in for this purpose, even if it works now wrong in your instance.

**Comment by TheWrongGuy** [comment]
Since this seems to be related to older conversations I  was wondering if there is a tool or something to make a database integrity / consistency check. Maybe one earlier migration gone wrong or something. Since the issue with wrong sidebar status was adressed in RC. 3.8.0 issue 19209 I was wondering if maybe a migration was skipped when I upgraded from 3.6.1 to 3.9.1 directly. Can a dev maybe tell me the migration number which was run upgrading to 3.8.0? (@sampaiodiego "like 179,rerun in my other issue?) Maybe I can try to rerun the migration.

**Comment by TheWrongGuy** [comment]
I did more testing and found something interesting using the REST API. I compared the results collected via the method rooms.info. (I erased some information for privacy reasons)

Direct messages room with broken user status:

{"room":{"_id":"1234567890123456789","usernames":["firstname1.lastname1","firstname2.lastname2"],"_updatedAt":"2020-12-14T14:00:44.793Z","t":"d","msgs":2344,"ts":"2018-11-22T14:02:17.198Z","usersCount":2,"lm":"2020-12-14T14:00:44.767Z","lastMessage":{"_id":"123129318313892","rid":"12472487294721847487","msg":"xxxxxxxxxxxxxxxxxxxxxxx","ts":"2020-12-14T14:00:44.767Z","u":{"_id":"dsdfasdhsdh1212","username":"firstname1.lastname1","name":"Name"},"_updatedAt":"2020-12-14T14:00:44.789Z","mentions":[],"channels":[]}},"success"

Direct messages room with working user status (newly created for testing purposes):

{"room":{"_id":"235823578237237523985732875357","t":"d","usernames":["firstname1.lastname1","firstname2.lastname2"],"usersCount":2,"msgs":3,"ts":"2020-12-09T09:33:19.944Z","uids":["192313123123123","21478124724427"],"default":false,"ro":false,"sysMes":true,"_updatedAt":"2020-12-09T10:46:51.084Z","lm":"2020-12-09T10:46:51.056Z","lastMessage":{"_id":"1237213213131313","rid":"235823578237237523985732875357","msg":"xxx","ts":"2020-12-09T10:46:51.056Z","u":{"_id":"23123123123123","username":"firstname2.lastname2","name":"Name"},"_updatedAt":"2020-12-09T10:46:51.079Z","mentions":[],"channels":[]}},"success":true}

The following section from the new chat seems to be entirely missing in the old broken chat:

,"uids":["192313123123123","21478124724427"],"default":false,"ro":false,"sysMes":true,"_updatedAt":"2020-12-09T10:46:51.084Z","lm":"2020-12-09T10:46:51.056Z",

Therefore in the old chat is no direct referral to the user ids who are participating. Only both usernames. Maybe the status is bound to the UID and therefore not properly shown in the room since the UIDs don't show up there? Is there maybe a way to edit this room info and add the UID to find out if this fixes this issue? Can a dev maybe see which migration maybe was missed and caused this inconsistency?

Thanks in advance and best regards.

**Comment by sampaiodiego** [comment]
thanks @TheWrongGuy looks like you found something.

so it is indeed still missing the `uids` property, I think that explains why in the UI it still shows the `@` symbol.

can you please confirm if the `_id` of the room that is missing the `uids` field is not composed by the `_id` of the users? for example, let's say the DM is between the following users:
```
{
  "_id": "abde1234",
  "username": "user1.lastname1"
}

{
  "_id": "xyz987",
  "username": "user2.lastname2"
}
```
the room of a DM from both users should look like this:
```
{
  _id: "abde1234xyz987",
  uids: [
    "abde1234",
    "xyz987",
  ],
  usernames: [
    "user1.lastname1",
    "user2.lastname2"
  ]
}
```
as you can see, the `_id` is *composed* by both user ids together..

**Comment by TheWrongGuy** [comment]
@sampaiodiego Thanks for the reply! Sorry for creating a misunderstanding by messing around with the ids in my example. I edited my comment before so now so it represents the real id compositions and should make the differnce visible better. But even for the DM with broken status the DM id IS indeed a composition of the correct user Ids. I just verified that. However the additional "uids: ["id1", "id2"] field is missing.

**Comment by sampaiodiego** [comment]
thanks @TheWrongGuy .. this is weird, the [migration 179](https://github.com/RocketChat/Rocket.Chat/blob/develop/server/startup/migrations/v179.js) should fix this issue, I can't why it is not fixing it..

can you please try again shutting rocket.chat down, and then start it up using the env var `MIGRATION_VERSION=179,rerun` ?

**Comment by TheWrongGuy** [comment]
Thank you very much @sampaiodiego I think this indeed fixes the issue, too now. The first time I tried this in the other thread the UI still was broken. This time after doing the migration again the "@" symbols still were there, so I thought it didn't work at first. But fortunately I sent a test message into a broken chat and the "@" symbol changed to the offline dot! This is an improvement. I can't say for sure that it is working now, since this was the test environment where everybody was offline of course, but I have a good feeling. I will try to do this on the productive environment ASAP. I will let you know! My thoughts on this issue it feels that after writing into the chat maybe the uid gets added to the rooms properties. I can't confirm this now since the REST API wasn't working properly in my test environment.

**Comment by sampaiodiego** [comment]
awesome.. that's good news.. you can try logging out and logging back in, this should clear the client cache and show an updated rooms list.. alternatively you can try logging in on an incognito window, it will have no cache and you should see the same thing =)

**Comment by TheWrongGuy** [comment]
Oh yes, I didn't think about this. This could of course be cache related. I will test this tomorrow in the productive environment. I think the Migration didn't cause any issues. The server seems to be running fine.

**Comment by TheWrongGuy** [comment]
@sampaiodiego 

Thank you very much again! This did the trick and fixed the status for all conversations. At one point the server skipped the 179 Migration apperantly. Whats a little confusing is that when I changed the env variable at the first restart of the RC service the migration didn't happen. I had to restart a second time. Also when removing the env variable again I had to restart twice to not get the 179 migration again. Logging out and logging in again indeed fixed the "@" symbols at once. This issue is fixed for me. Thanks again for the great support!
